### PR TITLE
Adjust for roslyn change on ref parameters and NotNullIfNotNull

### PR DESCRIPTION
**Pull request by jcouv**
Roslyn now uses the R-value/flow-state for `ref` parameters, instead of the declared type (L-value). (PR https://github.com/dotnet/roslyn/pull/47952)
This affected call to `EnsureInitializedCore`, which now needs explicit type parmaeters. 

Roslyn now enforced `NotNullIfNotNull` attribute within method bodies. This affected `CreateWrapperOfType`, which now needs a suppression.
Fixes `D:\repos\runtime\src\coreclr\src\System.Private.CoreLib\src\System\Runtime\InteropServices\Marshal.CoreCLR.cs(592,13): error CS8825: Return value must be non-null because parameter 'o' is non-null. [D:\repos\runtime\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]`

FYI, more fixes will be need to integrate a new version of roslyn, mostly due to `UnmanagedCallersOnly`:

 ```
D:\repos\runtime\src\coreclr\src\System.Private.CoreLib\src\System\Runtime\CompilerServices\CastHelpers.cs(140,32): error CS0030: Cannot convert type 'nuint' to 'System.Runtime.CompilerServices.CastHelpers.CastResult' [D:\repos\runtime\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
 
D:\repos\runtime\src\libraries\System.Private.CoreLib\src\System\Globalization\CalendarData.Icu.cs(428,59): error CS1503: Argument 1: cannot convert from '&method group' to 'delegate*<char*, IntPtr, void>' [D:\repos\runtime\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
D:\repos\runtime\src\libraries\System.Private.CoreLib\src\System\Globalization\CalendarData.Windows.cs(269,51): error CS1503: Argument 1: cannot convert from '&method group' to 'delegate*<char*, uint, IntPtr, void*, Interop.BOOL>' [D:\repos\runtime\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
D:\repos\runtime\src\libraries\System.Private.CoreLib\src\System\Globalization\CalendarData.Windows.cs(421,56): error CS8786: Calling convention of 'CalendarData.EnumCalendarsCallback(char*, uint, IntPtr, void*)' is not compatible with 'Default'. [D:\repos\runtime\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
D:\repos\runtime\src\libraries\System.Private.CoreLib\src\System\Globalization\CultureData.Nls.cs(120,55): error CS8786: Calling convention of 'CultureData.EnumSystemLocalesProc(char*, uint, void*)' is not compatible with 'Default'. [D:\repos\runtime\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
D:\repos\runtime\src\libraries\System.Private.CoreLib\src\System\Globalization\CultureData.Nls.cs(407,49): error CS8786: Calling convention of 'CultureData.EnumTimeCallback(char*, void*)' is not compatible with 'Default'. [D:\repos\runtime\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
D:\repos\runtime\src\libraries\System.Private.CoreLib\src\System\Globalization\CultureData.Nls.cs(493,55): error CS8786: Calling convention of 'CultureData.EnumAllSystemLocalesProc(char*, uint, void*)' is not compatible with 'Default'. [D:\repos\runtime\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
D:\repos\runtime\src\libraries\System.Private.CoreLib\src\System\Globalization\CultureData.Nls.cs(521,59): error CS8786: Calling convention of 'CultureData.EnumAllSystemLocalesProc(char*, uint, void*)' is not compatible with 'Default'. [D:\repos\runtime\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
```

### CODE REVIEW

File: src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.CoreCLR.cs [hunk 1]
```diff
@@ -585,7 +585,7 @@ public static bool SetComObjectData(object obj, object key, object? data)
                 if (!SetComObjectData(o, t, Wrapper))
                 {
                     // Another thead already cached the wrapper so use that one instead.
-                    Wrapper = GetComObjectData(o, t);
+                    Wrapper = GetComObjectData(o, t)!;
```
Comment (stephentoub) [inline src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.CoreCLR.cs:588]: Why is this needed?

Comment (stephentoub) [reply src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.CoreCLR.cs:588]: Oh, I see your pr explanation.

**Review by stephentoub — APPROVED** [review]

**Comment by stephentoub** [comment]
@danmosemsft, assuming we're going to take an updated compiler into the release branch for .NET 5, we'll need this change, plus whatever change addresses the unmanaged callers updates @jcouv mentions.

**Comment by danmosemsft** [comment]
@jeffschw where did we land on compiler updates? I think we have a principle that we build release with release so we must keep taking updates? Any matching fixes would likely be tell mode.

**Comment by am11** [comment]
> I think we have a principle that we build release with release so we must keep taking updates?

There was some related discussion about it in this thread for SDK 3.1 release, still using beta version of csc: https://github.com/dotnet/installer/issues/5857.

**Comment by jcouv** [comment]
@stephentoub I'm not sure the approval policy for `runtime` repo. If one approval suffices and there is no special ask-mode now, feel free to merge/squash. Thanks

**Comment by stephentoub** [comment]
Thanks, @jcouv.

**Comment by danmosemsft** [comment]
I noticed @ViktorHofer did https://github.com/dotnet/runtime/pull/42269 which I guess (?) means we will continue to get new compilers.

@ViktorHofer FYI that this change will need to get into 5.0 when that happens.

BTW, is there an `@` handle for the dotnet/runtime infra crew? I think that might be useful.

**Comment by ViktorHofer** [comment]
> I noticed @ViktorHofer did #42269 which I guess (?) means we will continue to get new compilers.

Depends on if Arcade will continue to update the compiler in their release/5.x branch (which doesn't exist yet).

> BTW, is there an @ handle for the dotnet/runtime infra crew? I think that might be useful.

`@dotnet/runtime-infrastructure`

**Comment by ViktorHofer** [comment]
> @ViktorHofer FYI that this change will need to get into 5.0 when that happens.

if not necessary for 5.0 RTW, we could/should port this into 5.0 for servicing readiness.
